gobblin-cluster/src/main/java/org/apache/gobblin/cluster/SingleTask.java

```diff
@@ -93,8 +109,17 @@ public void run()
         .createDefaultTopLevelBroker(jobConfig, GobblinScopeTypes.GLOBAL.defaultScopeInstance())) {
       SharedResourcesBroker<GobblinScopeTypes> jobBroker = getJobBroker(_jobState, globalBroker);
 
-      _taskAttempt = _taskAttemptBuilder.build(workUnits.iterator(), _jobId, _jobState, jobBroker);
-      _taskAttempt.runAndOptionallyCommitTaskAttempt(GobblinMultiTaskAttempt.CommitPolicy.IMMEDIATE);
+      // Secure atomicity of taskAttempt's execution.
+      // Signaling blocking threads if any whenever taskAttempt is nonNull.
+      _taskAttempt = _taskAttemptBuilder.build(getWorkUnits().iterator(), _jobId, _jobState, jobBroker);
+
+      _lock.lock();
+      try {
+        _taskAttemptBuilt.signal();
+        _taskAttempt.runAndOptionallyCommitTaskAttempt(GobblinMultiTaskAttempt.CommitPolicy.IMMEDIATE);
```

Comment:
Maybe I am missing something obvious - won't this block until the taskAttempt is finished? If so, is it right to hold the lock() until the attempt is finished? For instance, if a cancel() is invoked, won't the cancel() be waiting forever for the lock()? Alternately, if the cancel() acquires the lock() before the taskAttempt is built, then taskAttempt may be created after the cancel() fails.

Reply:
I was mistakenly thinking runAndOptionallyCommitTaskAttempt will return immediately. Then how about move this method out and make the build of taskAttemp inside the lock block?